### ISSUE TITLE
Wasserstein metric

### ISSUE THREAD

**Issue by farquasar1** [phillipi/pix2pix#47]
I would like to use the Wasserstein metric (https://github.com/martinarjovsky/WassersteinGAN/)  on this code. Any suggestion on how to implement it?

**Comment by junyanz** [comment]
Hi farquasar1, I just created a branch for [wgan-pix2pix](https://github.com/junyanz/pix2pix/tree/wgan). I haven't fully tested it. But I think it might help your implementation.

**Comment by farquasar1** [comment]
Thanks. It works but the in my case the L1 error is higher than without Wasserstein. I applying it to image segmentation

**Comment by farquasar1** [comment]
Just test your code but something is wrong as the output is producing the
same image as the input -  tested on two sets of images, seems to be
learning a segmentation of the input Image, not generating the output.
Trained from AtoB

**Comment by junyanz** [comment]
It's weird. Here is my training plot. The Wasserstein metric (the green line) does decrease over time. 
<img src='https://people.eecs.berkeley.edu/~junyanz/tmp4321/pix2pix/D_plot.png'>
In the branch, I provided a `train.sh` for training wgan. I recommend you (1) try the latest code of wgan branch on the facades dataset. (2) try the pix2pix code (with regular gan) on your own datasets.

**Comment by Quasimondo** [comment]
I gave the Wasserstein metric a try, but it appears to be extremely slow, something like 10 to 20 slower than the regular version. Is that expected?

**Comment by ppwwyyxx** [comment]
By slow do you mean:
1. more time for each iteration
2. more time to reach certain image quality
I saw 2 when I trained it.

**Comment by phillipi** [comment]
It is expected to be slower (per iteration) since the wgan code runs 5 iterations of the discriminator for every one iteration of the generator (see `opt.D_iters`). This follows the choice in Algorithm 1 from the [wgan paper](https://arxiv.org/pdf/1701.07875.pdf).

It would be interesting to try varying `opt.D_iters` and see how low a value you can get away with!

**Comment by Quasimondo** [comment]
Yes, slow in 1 and 2 I actually did not have the patience to wait until it got to a similar quality as the default metric in the same time (I had it training for a day or so on a training set of ~10000 items)

**Comment by royal-feng** [comment]
@junyanz I read your code in Wgan，I want to ask for why use the real_label set to -1 to backward in 259 line, the code is`netD:backward(real_AB, label)`, I only know train D is a regression problem, I don't know how to train it?

**Comment by rohitsaluja22** [comment]
@junyanz I cannot find your code for wganpix2pix. 
@all here is a good blog to convert pytorch vanilla code to wgan code..
https://lilianweng.github.io/lil-log/2017/08/20/from-GAN-to-WGAN.html

If you are able to convert any DCGAN code to convert one image to another to DC-WGAN, please let me know.

**Comment by weiqing1001** [comment]
> Hi farquasar1, I just created a branch for [wgan-pix2pix](https://github.com/junyanz/pix2pix/tree/wgan). I haven't fully tested it. But I think it might help your implementation.

Hello, the page is lost. Is there another page to find the branch. Thank you,

**Comment by egealpturkyenere** [comment]
> It's weird. Here is my training plot. The Wasserstein metric (the green line) does decrease over time. ![](https://camo.githubusercontent.com/03cc56479c8bcc5b7a8c30fcf5878964868588bdd0e5b6422bc0bbdb718817fb/68747470733a2f2f70656f706c652e656563732e6265726b656c65792e6564752f7e6a756e79616e7a2f746d70343332312f706978327069782f445f706c6f742e706e67) In the branch, I provided a `train.sh` for training wgan. I recommend you (1) try the latest code of wgan branch on the facades dataset. (2) try the pix2pix code (with regular gan) on your own datasets.

The page is lost. Is there somewhere that I can find the the branch?

**Comment by ALISAXXW** [comment]
您的邮件已收到，我会尽快回复~~~
